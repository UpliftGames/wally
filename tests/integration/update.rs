use crate::{temp_project::TempProject, util::read_path};
use fs_err as fs;
use insta::assert_snapshot;
use libwally::{
    lockfile::Lockfile,
    manifest::{Manifest, Package},
    package_id::PackageId,
    package_name::PackageName,
    package_req::PackageReq,
    Args, GlobalOptions, InstallSubcommand, PackageSpec, Subcommand, UpdateSubcommand,
};
use semver::Version;
use std::{
    convert::{TryFrom, TryInto},
    path::Path,
    str::FromStr,
};

#[test]
fn test_test() {
    let source_project = Path::new(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/test-projects/diamond-graph/root/0.1.0"
    ));

    let project = TempProject::new(&source_project).unwrap();
    run_download(&project);

    let args = Args {
        global: GlobalOptions {
            test_registry: true,
            ..Default::default()
        },
        subcommand: Subcommand::Update(UpdateSubcommand {
            project_path: project.path().to_owned(),
            package_specs: Vec::new(),
        }),
    };

    args.run().unwrap();
}

#[test]
fn generate_new_lockfile_if_missing() {
    let source_project = Path::new(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/test-projects/diamond-graph/root/0.1.0"
    ));

    let project = TempProject::new(&source_project).unwrap();

    let args = Args {
        global: GlobalOptions {
            test_registry: true,
            ..Default::default()
        },
        subcommand: Subcommand::Update(UpdateSubcommand {
            project_path: project.path().to_owned(),
            package_specs: Vec::new(),
        }),
    };

    let result = args.run();

    assert!(result.is_ok(), "It should of ran without any errors.");

    assert!(
        fs::metadata(project.path().join("wally.lock"))
            .unwrap()
            .is_file(),
        "It should've of generated a new wally.lock file."
    )
}

#[test]
fn do_nothing() {
    let source_project = Path::new(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/test-projects/diamond-graph/root/0.1.0"
    ));

    let project = TempProject::new(&source_project).unwrap();

    run_download(&project);

    Args {
        global: GlobalOptions {
            test_registry: true,
            ..Default::default()
        },
        subcommand: Subcommand::Update(UpdateSubcommand {
            project_path: project.path().to_owned(),
            package_specs: Vec::new(),
        }),
    }
    .run()
    .unwrap();

    let first_snapshot = read_path(project.path()).unwrap();

    run_update(&project);

    let second_snapshot = read_path(project.path()).unwrap();

    assert!(
        first_snapshot == second_snapshot,
        "Update is idiompotent assuming the same state."
    );
}

#[test]
fn update_all_dependencies() {
    let source_project = Path::new(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/test-projects/diamond-graph/root/0.1.0"
    ));

    let project = TempProject::new(&source_project).unwrap();

    let mut manifest = Manifest::load(project.path()).unwrap();

    run_download(&project);

    let first_lockfile = fs::read_to_string(project.path().join("wally.lock")).unwrap();
    assert_snapshot!(first_lockfile);

    manifest.server_dependencies.insert(
        "A".into(),
        PackageReq::from_str("diamond-graph/direct-dependency-a@0.1.0".into()).unwrap(),
    );

    Args {
        global: GlobalOptions {
            test_registry: true,
            ..Default::default()
        },
        subcommand: Subcommand::Update(UpdateSubcommand {
            project_path: project.path().to_owned(),
            package_specs: Vec::new(),
        }),
    }
    .run()
    .unwrap();

    let second_lockfile = fs::read_to_string(project.path().join("wally.lock")).unwrap();
    assert_snapshot!(second_lockfile);
}

#[test]
fn updated_named_dependency() {
    let source_project = Path::new(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/test-projects/diamond-graph/root/0.1.0"
    ));

    let project = TempProject::new(&source_project).unwrap();

    let mut manifest = Manifest::load(project.path()).unwrap();

    run_download(&project);

    let first_lockfile = fs::read_to_string(project.path().join("wally.lock")).unwrap();
    assert_snapshot!(first_lockfile);

    manifest.server_dependencies.insert(
        "A".into(),
        PackageReq::from_str("diamond-graph/direct-dependency-a@0.1.0".into()).unwrap(),
    );

    Args {
        global: GlobalOptions {
            test_registry: true,
            ..Default::default()
        },
        subcommand: Subcommand::Update(UpdateSubcommand {
            project_path: project.path().to_owned(),
            package_specs: vec![PackageSpec::Named(
                PackageName::new("diamond-graph", "direct-dependency-a").unwrap(),
            )],
        }),
    }
    .run()
    .unwrap();

    let second_lockfile = fs::read_to_string(project.path().join("wally.lock")).unwrap();
    assert_snapshot!(second_lockfile);
}

#[test]
fn update_named_dependency() {
    let source_project = Path::new(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/test-projects/diamond-graph/root/0.1.0"
    ));

    let project = TempProject::new(&source_project).unwrap();

    let mut manifest = Manifest::load(project.path()).unwrap();

    run_download(&project);

    let first_lockfile = fs::read_to_string(project.path().join("wally.lock")).unwrap();
    assert_snapshot!(first_lockfile);

    manifest.server_dependencies.insert(
        "A".into(),
        PackageReq::from_str("diamond-graph/direct-dependency-a@0.1.0".into()).unwrap(),
    );

    run_update_with_specs(
        &project,
        vec![PackageSpec::Required(
            PackageReq::from_str("diamond-graph/direct-dependency-a@0.1.0".into()).unwrap(),
        )],
    );

    let second_lockfile = fs::read_to_string(project.path().join("wally.lock")).unwrap();
    assert_snapshot!(second_lockfile);
}

fn update_list_of_specs() {}

fn update_should_download_new_dependences() {}

fn run_update(project: &TempProject) {
    run_update_with_specs(project, Vec::new())
}

fn run_update_with_specs(project: &TempProject, specs: Vec<PackageSpec>) {
    Args {
        global: GlobalOptions {
            test_registry: true,
            ..Default::default()
        },
        subcommand: Subcommand::Update(UpdateSubcommand {
            project_path: project.path().to_owned(),
            package_specs: specs,
        }),
    }
    .run()
    .unwrap();
}

fn run_download(project: &TempProject) {
    Args {
        global: GlobalOptions {
            test_registry: true,
            ..Default::default()
        },
        subcommand: Subcommand::Install(InstallSubcommand {
            project_path: project.path().to_owned(),
        }),
    }
    .run()
    .unwrap();
}
