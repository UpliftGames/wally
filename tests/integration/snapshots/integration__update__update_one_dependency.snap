---
source: tests/integration/update.rs
expression: result
---
ServerPackages:
  A.lua: "return require(script.Parent._Index[\"diamond-graph_direct-dependency-a@0.1.1\"][\"direct-dependency-a\"])\n"
  B.lua: "return require(script.Parent._Index[\"diamond-graph_direct-dependency-b@0.1.0\"][\"direct-dependency-b\"])\n"
  _Index:
    diamond-graph_direct-dependency-a@0.1.1:
      Indirect.lua: "return require(script.Parent.Parent[\"diamond-graph_indirect-dependency-a@0.1.1\"][\"indirect-dependency-a\"])\n"
      direct-dependency-a:
        default.project.json: "{\n\t\"name\": \"direct-dependency-a\",\n\t\"tree\": {\n\t\t\"Packages\": {\n\t\t\t\"$path\": \"Packages\"\n\t\t},\n\t\t\"$path\": \"src\"\n\t}\n}"
        src:
          init.lua: "local Indirect = require(script.Parent.Indirect)\n\nreturn if Indirect == 420 then \"Bingo!\" else \"Naw...\""
        wally.toml: "[package]\nname = \"diamond-graph/direct-dependency-a\"\nversion = \"0.1.1\"\nlicense = \"MIT\"\nrealm = \"server\"\nregistry = \"test-registries/primary-registry\"\n\n[server-dependencies]\nIndirect = \"diamond-graph/indirect-dependency-a@0.1.1\""
    diamond-graph_direct-dependency-b@0.1.0:
      Indirect.lua: "return require(script.Parent.Parent[\"diamond-graph_indirect-dependency-a@0.2.0\"][\"indirect-dependency-a\"])\n"
      direct-dependency-b:
        default.project.json: "{\n\t\"name\": \"direct-dependency-b\",\n\t\"tree\": {\n\t\t\"Packages\": {\n\t\t\t\"$path\": \"Packages\"\n\t\t},\n\t\t\"$path\": \"src\"\n\t}\n}"
        src:
          init.lua: "local Indirect = require(script.Parent.Indirect)\n\nreturn Indirect == 420"
        wally.toml: "[package]\nname = \"diamond-graph/direct-dependency-b\"\nversion = \"0.1.0\"\nlicense = \"MIT\"\nrealm = \"server\"\nregistry = \"test-registries/primary-registry\"\n\n[server-dependencies]\nIndirect = \"diamond-graph/indirect-dependency-a@0.2.0\""
    diamond-graph_indirect-dependency-a@0.1.1:
      indirect-dependency-a:
        default.project.json: "{\n\t\"name\": \"indirect-dependency-a\",\n\t\"tree\": {\n\t\t\"$path\": \"src\"\n\t}\n}"
        src:
          init.lua: "-- Zesty\nreturn function()\n\treturn 420\nend\n"
        wally.toml: "[package]\nname = \"diamond-graph/indirect-dependency-a\"\nversion = \"0.1.1\"\nlicense = \"MIT\"\nrealm = \"server\"\nregistry = \"test-registries/primary-registry\"\n"
    diamond-graph_indirect-dependency-a@0.2.0:
      indirect-dependency-a:
        default.project.json: "{\n\t\"name\": \"indirect-dependency-a\",\n\t\"tree\": {\n\t\t\"$path\": \"src\"\n\t}\n}"
        src:
          init.lua: "return function()\n\treturn 42\nend\n"
        wally.toml: "[package]\nname = \"diamond-graph/indirect-dependency-a\"\nversion = \"0.2.0\"\nlicense = \"MIT\"\nrealm = \"server\"\nregistry = \"test-registries/primary-registry\"\n"
default.project.json: "{\n\t\"name\": \"root\",\n\t\"tree\": {\n\t\t\"Packages\": {\n\t\t\t\"$path\": \"Packages\"\n\t\t},\n\t\t\"$path\": \"src\"\n\t}\n}"
src:
  init.lua: "local A = require(script.Parent.A)\nlocal B = require(script.Parent.B)\n\nreturn `{A} {B}`"
wally.lock: "# This file is automatically @generated by Wally.\n# It is not intended for manual editing.\nregistry = \"test\"\n\n[[package]]\nname = \"diamond-graph/direct-dependency-a\"\nversion = \"0.1.1\"\ndependencies = [[\"Indirect\", \"diamond-graph/indirect-dependency-a@0.1.1\"]]\n\n[[package]]\nname = \"diamond-graph/direct-dependency-b\"\nversion = \"0.1.0\"\ndependencies = [[\"Indirect\", \"diamond-graph/indirect-dependency-a@0.2.0\"]]\n\n[[package]]\nname = \"diamond-graph/indirect-dependency-a\"\nversion = \"0.1.1\"\ndependencies = []\n\n[[package]]\nname = \"diamond-graph/indirect-dependency-a\"\nversion = \"0.2.0\"\ndependencies = []\n\n[[package]]\nname = \"diamond-graph/root\"\nversion = \"0.1.0\"\ndependencies = [[\"A\", \"diamond-graph/direct-dependency-a@0.1.1\"], [\"B\", \"diamond-graph/direct-dependency-b@0.1.0\"]]\n"
wally.toml: "[package]\nname = \"diamond-graph/root\"\nversion = \"0.1.0\"\nlicense = \"MIT\"\nrealm = \"server\"\nregistry = \"test-registries/primary-registry\"\n\n[server-dependencies]\nA = \"diamond-graph/direct-dependency-a@0.1.0\"\nB = \"diamond-graph/direct-dependency-b@0.1.0\""

